proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,537 @@
+* Implementation: [apple/swift#34272](https://github.com/apple/swift/pull/34272)
+
+
+## Introduction
+
+Property Wrappers were [introduced in Swift 5.1](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md), and have since become a popular feature abstracting away common accessor patterns for properties. Currently, applying a property wrapper is solely permitted at local and type context. However, with increasing adoption, demand for extending _where_ property wrappers can be applied has emerged. This proposal aims to extend property wrappers to function and closure parameters.
+
+
+## Motivation
+
+Property wrappers have undoubtably been very successful. Applying a property wrapper to a property is enabled by an incredibly lightweight and expressive syntax. Therefore, library authors can expose complex behavior through easily understandable property-wrapper types in an efficient manner. For instance, frameworks such as [SwiftUI](https://developer.apple.com/documentation/swiftui/) and [Combine](https://developer.apple.com/documentation/combine) introduce property wrappers such as [`State`](https://developer.apple.com/documentation/swiftui/state), [`Binding`](https://developer.apple.com/documentation/swiftui/binding) and [`Published`](https://developer.apple.com/documentation/combine/published) respectively, to expose elaborate behavior through a succinct interface, helping craft expressive yet simple APIs. However, property wrappers are only applicable to local variables and type properties, shattering the illusion that they helped realize in the first place when working with parameters.
+
+### Memberwise initialization
+
+Currently, property wrappers on stored struct properties already interact with parameters through the generated memberwise initializer for the struct. However, property wrapper attributes are not supported on parameters, which leads to really complicated and nuanced rules for which type the generated memberwise initializer should accept.
+
+The compiler will choose the wrapped value type to offer a convenience to the call-site when the property wrapper supports initalization through the wrapped value type via `init(wrappedValue:)`:
+
+```swift
+import SwiftUI
+
+struct TextEditor {
+  @State var document: Optional<URL>
+}
+
+func openEditor(with swiftFile: URL) -> TextEditor {
+  TextEditor(document: swiftFile)
+}
+```
+
+However, this can take flexibility away from the call-site if the property wrapper has other `init` overloads, because the call-site can cannot choose a different initializer. Further, if the property wrapper is default initialized in the struct, then the memberwise initializer will choose the backing wrapper type, even if the wrapper supports `init(wrappedValue:)`. This creates unnecessary boilerplate at call-sites that do want to use `init(wrappedValue:)`:
+
+```swift
+import SwiftUI
+
+struct TextEditor {
+  @State() var document: Optional<URL>
```

Comment:
Are the parenthesis right after `@State` intentional? If yes, I'm kind of lost as to why; if there _is_ a reason, maybe we could add a comment right below it to explain why the paranthesis are there.

Reply:
Yes, empty parenthesis after the wrapper attribute is the way to do default initialization of the backing wrapper. I thought this was clear from the sentence above:

> Further, if the property wrapper is default initialized in the struct ...

Reply:
Maybe we could change it to: 

> Further, if the property wrapper is default initialized in the struct – with an `init()` overload – ...

I say that because "default initialized" may not instinctively lead the reader to assume that we mean the plain `State` initializer: `init()`; not to mention, that this isn't the form we usually see `State` initialized in the documentation.

Reply:
Whoops, I was mixing up my terminology. This is explicit initialization via `init()`. Fixed in the description above the example.